source/agora/common/BlockStorage.d

```diff
@@ -793,3 +793,263 @@ public size_t readSizeType (File file) @trusted
+
+    /// Index is block hash
+    private IndexHash hash_idx;
+
+    /// Ctor
+    public this ()
+    {
+        this.height_idx = new IndexHeight();
+        this.hash_idx = new IndexHash();
+    }
+
+    /***************************************************************************
+
+        Read the last block from array.
+
+        Params:
+            block = will contain the block if the read was successful
+
+        Returns:
+            Returns true if success, otherwise returns false.
+
+    ***************************************************************************/
+
+    public bool readLastBlock (ref Block block) @safe nothrow
+    {
+        if (this.height_idx.length == 0)
+            return false;
+
+        return this.readBlock(block, this.height_idx.back.position);
+    }
+
+    /***************************************************************************
+
+        Save block to array.
+
+        Params:
+            block = `Block` to save
+
+        Returns:
+            Returns true if success, otherwise returns false.
+
+    ***************************************************************************/
+
+    public bool saveBlock (const ref Block block) @safe nothrow
+    {
+        try
+        {
+            if (this.blocks.length != block.header.height)
+                return false;
+
+            size_t block_position = this.blocks.length;
+
+            this.blocks ~= block;
+
+            // add to index of heigth
+            this.height_idx.insert(
+                HeightPosition(
+                    block.header.height,
+                    block_position
+                )
+            );
+
+            // add to index of hash
+            ubyte[Hash.sizeof] hash_bytes = hashFull(block.header)[];
+            this.hash_idx.insert(
+                HashPosition(
+                    hash_bytes,
+                    block_position
+                )
+            );
+            return true;
+        }
+        catch (Exception ex)
+        {
+            this.writeLog("MemBlockStorage.saveBlock: ", ex);
+            return false;
+        }
+    }
+
+    /***************************************************************************
+
+        Read block from array.
+
+        Params:
+            block = `Block` to read
+            height = height of `Block`
+
+        Returns:
+            Returns true if success, otherwise returns false.
+
+    ***************************************************************************/
+
+    public bool readBlock (ref Block block, size_t height) @safe nothrow
+    {
+        if ((this.height_idx.length == 0) ||
+            (this.height_idx.back.height < height))
+            return false;
+
+        auto finds
+            = this.height_idx[].find!( (a, b) => a.height == b)(height);
+
+        if (finds.empty)
+            return false;
+
+        try
+        {
+            () @trusted {
+                block = cast(Block)this.blocks[finds.front.position];
```

Comment:
This is the problematic part, because it's casting away const.

However I have an idea how to solve it. Using serialization.

Reply:
@Geod24 this code is in `MemBlockStorage`, which is used in unittests. 

I only see two options here:

- Instead of storing `const(Block)[]`, store a serialized `ubyte[]` and deserialize the block here.
- Do a deep copy of the struct to get a mutable Block

Can you think of another solution?

Reply:
It's a good solution. I checked the review and have already fixed it.

Reply:
What happened to the code which fixed the cast? It seems to have disappeared.

Reply:
Indeed this cast is problematic.